### PR TITLE
[WIP] Initialise audio separation task and backbone.py

### LINKED ISSUE #548: Add an AUDIO_SOURCE_SEPARATION_BACKBONES FlashRegistry with a few simple models from Asteroid

### PR DESCRIPTION
**Pull request by saswatpp**
## What does this PR do?
This PR initialises directories for Audio Source Separation and adds backbone.py file 
(Fixes #548) to register the backbones from https://github.com/asteroid-team/asteroid/tree/master/asteroid/models. 
Currently I have added only `ConvTasNet`, `DPRNNTasNet` to the registry and will extend it once the overall structure is confirmed. 

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by saswatpp** [comment]
@ethanwharris, the doubt is that asteroid uses the following design to add a pretrained model 
```python
from asteroid.models import ConvTasNet
model = ConvTasNet.from_pretrained('mpariente/ConvTasNet_WHAM!_sepclean')
```

and we require something like the following to include pretrained models for

```python
AudioSourceSeparator(backbone="ConvTasNet", n_src=datamodule.n_src, pretrained=True)
```

So how would we implement this ? do we have to maintain a dictionary of pretrained model names (like `'mpariente/ConvTasNet_WHAM!_sepclean'`) as above ? suggest. 

Thank you 😃

File: flash/audio/source_separation/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+from flash.audio.source_separation.data import AudioSourceSeparationData
+from flash.audio.source_separation.model import AudioSourceSeparator
```
Comment (tchaton) [inline flash/audio/source_separation/__init__.py:2]: Should we simplify it for `AudioSeparator` ?

Comment (saswatpp) [reply flash/audio/source_separation/__init__.py:2]: Maybe just `SourceSeparator` . Because its already inside the audio domain ?

Comment (tchaton) [reply flash/audio/source_separation/__init__.py:2]: For consistency, we keep the data type in the task name as prefix.

noob question: Mind explaining me what source mean in this context. I believe it could be confusing for new users.

**Comment by tchaton** [comment]
> @ethanwharris, the doubt is that `asteroid` uses the following design to add a pretrained model
> 
> ```python
> from asteroid.models import ConvTasNet
> model = ConvTasNet.from_pretrained('mpariente/ConvTasNet_WHAM!_sepclean')
> ```
> 
> and we require something like the following to include pretrained models for
> 
> ```python
> AudioSourceSeparator(backbone="ConvTasNet", n_src=datamodule.n_src, pretrained=True)
> ```
> 
> So how would we implement this ? do we have to maintain a dictionary of pretrained model names (like `'mpariente/ConvTasNet_WHAM!_sepclean'`) as above ? please suggest.
> 
> Thank you 😃

Dear @saswatpp,

Have a look there for inspirations: https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/pointcloud/segmentation/open3d_ml/backbones.py#L67

@ethanwharris What are your thoughts about the following:

We could have something like this

```python
# shoudn't be pretrained. If pretrained=True is passed, it should fail and propose its pretrained version.
AudioSourceSeparator(backbone="convtasnet", n_src=datamodule.n_src)
```


```python
# shoudn't be pretrained. If pretrained=True is passed, it should fail and propose its pretrained version.
AudioSourceSeparator(backbone="convtasnet", n_src=datamodule.n_src)
```

```python
@register
def convtasknet_wham():
       return  ConvTasNet.from_pretrained('mpariente/ConvTasNet_WHAM!_sepclean')
```

```python
AudioSourceSeparator(backbone="convtasknet_wham", n_src=datamodule.n_src)
```

and we require something like the following to include pretrained models for

```python
AudioSourceSeparator(backbone="ConvTasNet", n_src=datamodule.n_src, pretrained=True)
```

File: flash/audio/source_separation/__init__.py (hunk 1 above)
Comment (saswatpp) [reply flash/audio/source_separation/__init__.py:2]: Yes I understand. `Source` is analogous to `Speaker` in the an audio file which has lets say multiple speakers speaking simultaneously. So the speech for multiple speakers are isolated for individual speakers and separated.

**Comment by saswatpp** [comment]
I have added separate Backbone registry for pretrained models and randomly initialised models. The backbones contains the name of the dataset, just like @tchaton suggested but there are numerous pretrained models for different `n_src`(the number of masks generated) and task on which the model was trained : `sep_clean` or `sep_noisy` and `enhsingle`(single source enhacement).

Along with that there are no common naming schemes of the pretrained model and each author follows their own style, hence I did not automate the link generation process and manually added each link (link here means the path from where the models are fetched). For example : `JorisCos/ConvTasNet_Libri3Mix_sepnoisy_16k` is one of the model sources where the `n_src` is `3`, `sep_noisy` is task and `LibriMix` is dataset. And some other author named the model link as `julien-c/DPRNNTasNet-ks16_WHAM_sepclean`. 

Can anyone provide any other ideas ? Thank you

File: flash/audio/source_separation/backbones.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    import asteroid.models as ast
+    from asteroid.models import BaseModel
+
+    ASTEROID_BACKBONE_CLASS = [ast.ConvTasNet, ast.DPRNNTasNet]
+    ASTEROID_CLASSES = {a.__name__.lower(): a for a in ASTEROID_BACKBONE_CLASS}
+
+    def _load_asteroid_model(
+        backbone: str,
+        n_src: int = None,
+        pretrained: bool = False,
```
Comment (tchaton) [inline flash/audio/source_separation/backbones.py:34]: How should we use pretrained ?

File: flash/audio/source_separation/backbones.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        if backbone not in ASTEROID_BACKBONES:
+            raise NotImplementedError(f"{backbone} is not implemented! Supported heads -> {ASTEROID_BACKBONES.keys()}")
+
+        return BaseModel.from_pretrained(url), n_src
+
+    ASTEROID_BACKBONES(
+        partial(
+            _load_pretrained_model,
+            backbone="convtasnet_libri1mix_enhsingle_16k",
+            n_src=1,
```
Comment (tchaton) [inline flash/audio/source_separation/backbones.py:72]: I am bit confused by `n_src=1`. Might explaining what it does ?

**Comment by tchaton** [comment]
Dear @saswatpp,

Yes, I like the pretrained model nomenclature.

We can make it even more explicit by taking n_src out as follows:

```py
ConvTasNet_LibriMix_3_sepnoisy_16k
# {MODEL_NAME}_{DATASET_NAME}_{N_SRC}_{TYPE}_{FREQUENCY}

```

We could organise the backbones and make them filterable as follows:

```py
AudioSourceSeparation.available_backbones(dataset, nr_src, type, frequency)
``` 

Best,
B.C

**Comment by ethanwharris** [comment]
Closing this as development of the AudioSourceSeparation task has gone stale
